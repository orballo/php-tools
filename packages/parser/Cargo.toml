[package]
name = "parser"
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true

[dependencies]
rowan.workspace = true
serde.workspace = true
serde_json.workspace = true
nom.workspace = true
lexer.path = "../lexer"
lexer_derive.path = "../lexer_derive"

[dev-dependencies]
insta.workspace = true

[profile.dev.package]
insta.opt-level = 3
similar.opt-level = 3
