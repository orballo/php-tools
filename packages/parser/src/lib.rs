pub mod php_parser;

pub use php_parser::{PHPParser, Rule as PHPRule};
