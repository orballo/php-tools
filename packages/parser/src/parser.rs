// @TODO: the parser core implementation should parse
// the pieces of the scripts, and once the script sections
// are identified, these are the ones that should be parsed
// with the extensions
//
// The text sections could potentially accept extensions to parse HTML, etc.

pub struct Parser {}

impl Parser {
    pub fn builder() -> ParserBuilder {
        ParserBuilder::new()
    }
}

// @TODO: We need to implement a Token struct.
// This struct is going to be used to store token kinds
// with their name, span, and &str to source.
//
// We need to do the lexing first, and if possible
// concurrently over multiple files.
//
// Parsing should be done after lexing, and only using
// the tokens from the lexer's output.
//
// Each feature/token of the language will be
// in a package called extension.
//
// The individual extensions can implement:
// - Token definition and parser (Optional)
// - Expresion definition and parser (Optional)
// - Statement definition and parser (Optional)
// - Bytecode instruction definition (Optional)
// - Bytecode instruction parser (Optional)
//
pub struct ParserBuilder {
    tokens: Vec<Token>,
    parsers: Vec<Box<dyn SubParser>>,
}

impl ParserBuilder {
    pub fn new() -> Self {
        Self {
            tokens: Vec::new(),
            parsers: Vec::new(),
        }
    }

    pub fn parser() -> Self {
        todo!()
    }

    pub fn parsers() -> Self {
        todo!()
    }
}

pub trait SubParser {}
