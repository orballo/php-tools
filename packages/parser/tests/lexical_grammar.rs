use insta::{assert_debug_snapshot, glob};
use parser::{PHPParser, PHPRule};
use pest::Parser;
use std::fs;

#[test]
fn lexical_grammar_snapshots() {
    let mut settings = insta::Settings::clone_current();
    settings.set_prepend_module_to_snapshot(false);

    settings.bind(|| {
        glob!("fixtures/*.php", |path| {
            let file = fs::read_to_string(path).unwrap();
            let mut parsed_file = PHPParser::parse(PHPRule::SCRIPT, &file)
                .expect("Something went wrong parsing the file");

            assert_debug_snapshot!(parsed_file.next().unwrap());
        })
    });
}
