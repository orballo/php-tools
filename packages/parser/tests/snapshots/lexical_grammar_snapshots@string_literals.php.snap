---
source: packages/parser/tests/lexical_grammar.rs
expression: parsed_file.next().unwrap()
input_file: packages/parser/tests/fixtures/string_literals.php
---
Pair {
    rule: SCRIPT,
    span: Span {
        str: "<?php\n\n// Single Quoted String Literals\n\n'This is a single quoted string literal';\n'This is a single quoted string literal with an escaped single quote: \\'';\n'This is a single quoted string literal with an escaped backslash: \\\\';\n'This is a single quoted string literal with a newline: \\n';\nb'This is a single quoted binary string literal';\n\n// Double Quoted String Literals\n\n\"This is a double quoted string literal\";\n\"This is a double quoted string literal with simple escape sequences: \\\\ \\\" \\$ \\e \\f \\n \\r \\t \\v\";\n\"This is a double quoted string literal with hoctal escape sequences: \\7 \\77 \\777\";\n\"This is a double quoted string literal with hexadecimal escape sequences: \\x8 \\xFF\";\n\"This is a double quoted string literal with unicode escape sequences: \\u{91223} \\u{00CCFF} \\u{1}\";\n\"This is a double quoted string literal with a literal backslash: \\bar\";\n\"This is a double quoted string literal with variable interpolation: $foo\";\nb\"This is a double quoted binary string literal\";\n",
        start: 0,
        end: 986,
    },
    inner: [
        Pair {
            rule: SCRIPT_SECTION,
            span: Span {
                str: "<?php\n\n// Single Quoted String Literals\n\n'This is a single quoted string literal';\n'This is a single quoted string literal with an escaped single quote: \\'';\n'This is a single quoted string literal with an escaped backslash: \\\\';\n'This is a single quoted string literal with a newline: \\n';\nb'This is a single quoted binary string literal';\n\n// Double Quoted String Literals\n\n\"This is a double quoted string literal\";\n\"This is a double quoted string literal with simple escape sequences: \\\\ \\\" \\$ \\e \\f \\n \\r \\t \\v\";\n\"This is a double quoted string literal with hoctal escape sequences: \\7 \\77 \\777\";\n\"This is a double quoted string literal with hexadecimal escape sequences: \\x8 \\xFF\";\n\"This is a double quoted string literal with unicode escape sequences: \\u{91223} \\u{00CCFF} \\u{1}\";\n\"This is a double quoted string literal with a literal backslash: \\bar\";\n\"This is a double quoted string literal with variable interpolation: $foo\";\nb\"This is a double quoted binary string literal\";\n",
                start: 0,
                end: 986,
            },
            inner: [
                Pair {
                    rule: OPEN_TAG,
                    span: Span {
                        str: "<?php",
                        start: 0,
                        end: 5,
                    },
                    inner: [],
                },
                Pair {
                    rule: SINGLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "'This is a single quoted string literal'",
                        start: 41,
                        end: 81,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 81,
                        end: 82,
                    },
                    inner: [],
                },
                Pair {
                    rule: SINGLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "'This is a single quoted string literal with an escaped single quote: \\''",
                        start: 83,
                        end: 156,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 156,
                        end: 157,
                    },
                    inner: [],
                },
                Pair {
                    rule: SINGLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "'This is a single quoted string literal with an escaped backslash: \\\\'",
                        start: 158,
                        end: 228,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 228,
                        end: 229,
                    },
                    inner: [],
                },
                Pair {
                    rule: SINGLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "'This is a single quoted string literal with a newline: \\n'",
                        start: 230,
                        end: 289,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 289,
                        end: 290,
                    },
                    inner: [],
                },
                Pair {
                    rule: SINGLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "b'This is a single quoted binary string literal'",
                        start: 291,
                        end: 339,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 339,
                        end: 340,
                    },
                    inner: [],
                },
                Pair {
                    rule: DOUBLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "\"This is a double quoted string literal\"",
                        start: 376,
                        end: 416,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 416,
                        end: 417,
                    },
                    inner: [],
                },
                Pair {
                    rule: DOUBLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "\"This is a double quoted string literal with simple escape sequences: \\\\ \\\" \\$ \\e \\f \\n \\r \\t \\v\"",
                        start: 418,
                        end: 515,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 515,
                        end: 516,
                    },
                    inner: [],
                },
                Pair {
                    rule: DOUBLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "\"This is a double quoted string literal with hoctal escape sequences: \\7 \\77 \\777\"",
                        start: 517,
                        end: 599,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 599,
                        end: 600,
                    },
                    inner: [],
                },
                Pair {
                    rule: DOUBLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "\"This is a double quoted string literal with hexadecimal escape sequences: \\x8 \\xFF\"",
                        start: 601,
                        end: 685,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 685,
                        end: 686,
                    },
                    inner: [],
                },
                Pair {
                    rule: DOUBLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "\"This is a double quoted string literal with unicode escape sequences: \\u{91223} \\u{00CCFF} \\u{1}\"",
                        start: 687,
                        end: 785,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 785,
                        end: 786,
                    },
                    inner: [],
                },
                Pair {
                    rule: DOUBLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "\"This is a double quoted string literal with a literal backslash: \\bar\"",
                        start: 787,
                        end: 858,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 858,
                        end: 859,
                    },
                    inner: [],
                },
                Pair {
                    rule: DOUBLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "\"This is a double quoted string literal with variable interpolation: $foo\"",
                        start: 860,
                        end: 934,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 934,
                        end: 935,
                    },
                    inner: [],
                },
                Pair {
                    rule: DOUBLE_QUOTED_STRING_LITERAL,
                    span: Span {
                        str: "b\"This is a double quoted binary string literal\"",
                        start: 936,
                        end: 984,
                    },
                    inner: [],
                },
                Pair {
                    rule: SEMICOLON,
                    span: Span {
                        str: ";",
                        start: 984,
                        end: 985,
                    },
                    inner: [],
                },
            ],
        },
    ],
}
